server/image_pull.go

```diff
@@ -80,16 +85,70 @@ func (s *Server) PullImage(ctx context.Context, req *pb.PullImageRequest) (resp
 			} else if tmpImgConfigDigest.String() == storedImage.ConfigDigest.String() {
 				log.Debugf(ctx, "image %s already in store, skipping pull", img)
 				pulled = img
+
+				// Skipped bytes metrics
+				if storedImage.Size != nil {
+					counter, err := metrics.CRIOImagePullsByNameSkipped.GetMetricWithLabelValues(img)
+					if err != nil {
+						log.Warnf(ctx, "Unable to write image pull name (skipped) metrics: %v", err)
+					} else {
+						counter.Add(float64(*storedImage.Size))
+					}
+				}
+
 				break
 			}
 			log.Debugf(ctx, "image in store has different ID, re-pulling %s", img)
 		}
 
-		_, err = s.StorageImageServer().PullImage(s.systemContext, img, &copy.Options{
-			SourceCtx:        &sourceCtx,
-			DestinationCtx:   s.systemContext,
-			OciDecryptConfig: dcc,
-		})
+		// Pull by collecting progress metrics
+		progress := make(chan types.ProgressProperties)
+		go func() {
+			_, err := s.StorageImageServer().PullImage(s.systemContext, img, &copy.Options{
+				SourceCtx:        &sourceCtx,
+				DestinationCtx:   s.systemContext,
+				OciDecryptConfig: dcc,
+				ProgressInterval: time.Second,
+				Progress:         progress,
+			})
+			pullError <- err
+		}()
+		go func() {
+			for p := range progress {
+				if p.Artifact.Size > 0 {
+					log.Debugf(ctx, "ImagePull (%v): %s (%s): %v bytes (%.2f%%)",
+						p.Event, img, p.Artifact.Digest, p.Offset,
+						float64(p.Offset)/float64(p.Artifact.Size)*100,
+					)
+				} else {
+					log.Debugf(ctx, "ImagePull (%v): %s (%s): %v",
```

Comment:
(Non-blocking: Maybe `%v bytes` to be the same as the other case)

Reply:
Fixed that as well.